### PR TITLE
Feature Relation item

### PR DESCRIPTION
**Pull request by iiLubos**
image
image

```
MMTextBubblesEditor {
          title: "MMTextAreaEditor"
          featuresModel: ListModel { ... }
}
```

### CODE REVIEW

File: app/qml/inputs/MMTextBubblesEditor.qml [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+MMAbstractEditor {
+  id: root
+
+  property var parentValue: parent.value ?? ""
+  property bool parentValueIsNull: parent.valueIsNull ?? false
+  property bool isReadOnly: parent.readOnly ?? false
+
+  readonly property real itemHeight: 36 * __dp
+  readonly property int rows: 3
+  required property ListModel featuresModel
```
Comment (tomasMizera) [inline app/qml/inputs/MMTextBubblesEditor.qml:24]: Please encapsulate private properties used only in one file in a separate `QtObject` at the end of the file to make the file more readable https://github.com/MerginMaps/mobile/blob/b3e30023de8a75a09deaa246ead1b5ec2f2a89db/docs/code_convention.md?plain=1#L151

File: app/qml/inputs/MMTextBubblesEditor.qml [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        id: footer
+
+        width: 100 * __dp
+        height: root.itemHeight
+        radius: 8 * __dp
+        color: __style.lightGreenColor
+
+        Text {
+          id: moreText
+
```
Comment (tomasMizera) [inline app/qml/inputs/MMTextBubblesEditor.qml:117]: How about binding the text here, won't that work?
```suggestion
  text: "+" + repeater.invisibleIds.length + qsTr(" more")
```

File: app/qml/inputs/MMTextBubblesEditor.qml [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+  readonly property int rows: 3
+  required property ListModel featuresModel
+
+  signal editorValueChanged( var newValue, var isNull )
+  signal openLinkedFeature( var linkedFeature )
+  signal createLinkedFeature( var parentFeature, var relation )
+
+  contentItemHeight: itemHeight * rows + 2 * flow.spacing + 20 * __dp
+
+  Component.onCompleted: root.hideFeaturesIfNeeded()
```
Comment (tomasMizera) [inline app/qml/inputs/MMTextBubblesEditor.qml:32]:
```suggestion
  Component.onCompleted: root.recalculate()
```

File: app/qml/inputs/MMTextBubblesEditor.qml [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+          MouseArea {
+            anchors.fill: parent
+            onClicked: root.openLinkedFeature( model.FeaturePair )
+          }
+        }
+      }
+
+      Rectangle {
+        id: footer
+
```
Comment (tomasMizera) [inline app/qml/inputs/MMTextBubblesEditor.qml:109]: Aren't you missing something like:
```qml
    visible: repeater.invisibleIds > 0
```
here? To hide this item if there is enough space to show all the items

File: app/qml/inputs/MMTextBubblesEditor.qml [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        delegate: Rectangle {
+          width: text.contentWidth + 24 * __dp
+          height: root.itemHeight
+          radius: 8 * __dp
+          color: __style.mediumGreenColor
+          visible: {
+            let isVisible = (y < 2 * root.itemHeight || (y < 3 * root.itemHeight && x + width < flow.width - footer.width - flow.spacing))
+            if(!isVisible && !repeater.invisibleIds.includes(model.FeatureId))
+              repeater.invisibleIds.push(model.FeatureId)
```
Comment (tomasMizera) [inline app/qml/inputs/MMTextBubblesEditor.qml:83]: Please see code convention about  js code: https://github.com/MerginMaps/mobile/blob/b3e30023de8a75a09deaa246ead1b5ec2f2a89db/docs/code_convention.md?plain=1#L164

```suggestion
            if (!isVisible && !repeater.invisibleIds.includes(model.FeatureId)) {
              repeater.invisibleIds.push(model.FeatureId)
            }
```

File: app/qml/inputs/MMTextBubblesEditor.qml [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+            footer.visible = repeater.invisibleIds.length > 0
+            moreText.text = ("+" + repeater.invisibleIds.length + qsTr(" more"))
```
Comment (tomasMizera) [inline app/qml/inputs/MMTextBubblesEditor.qml:85]: I hope both of these lines could be removed and moved as properties into the `footer`

File: app/qml/inputs/MMTextBubblesEditor.qml [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+            listLoader.focus = true
+          }
+        }
+      }
+    }
+  }
+
+  function hideFeaturesIfNeeded() {
+    repeater.invisibleIds = []
+    repeater.model = undefined
```
Comment (tomasMizera) [inline app/qml/inputs/MMTextBubblesEditor.qml:140]: It was `null` previously, not sure if it makes a difference

```suggestion
    repeater.model = null
```

File: app/qml/components/MMLinkedFeaturesDrawer.qml [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+import QtQuick
+import QtQuick.Controls
+import QtQuick.Controls.Basic
+import "../inputs"
+
+Drawer {
```
Comment (tomasMizera) [inline app/qml/components/MMLinkedFeaturesDrawer.qml:15]: Quick question about drawers, can there be multiple of them opened at the same time? This page can be opened multiple times at the same time

Comment (iiLubos) [reply app/qml/components/MMLinkedFeaturesDrawer.qml:15]: Yes, multiple drawers can be opened at the same time.

Comment (iiLubos) [reply app/qml/components/MMLinkedFeaturesDrawer.qml:15]: https://github.com/MerginMaps/mobile/assets/72412563/9ebb4d50-93b3-419f-ae11-64e96df5b829

Comment (iiLubos) [reply app/qml/components/MMLinkedFeaturesDrawer.qml:15]: Tested also on mobile

File: app/qml/inputs/MMTextBubblesEditor.qml (hunk 7 above)
Comment (iiLubos) [reply app/qml/inputs/MMTextBubblesEditor.qml:140]: It looks, behavior is the same

File: app/qml/inputs/MMTextBubblesEditor.qml (hunk 2 above)
Comment (iiLubos) [reply app/qml/inputs/MMTextBubblesEditor.qml:117]: unfortunately no, here the length is still 0

File: app/qml/inputs/MMTextBubblesEditor.qml (hunk 6 above)
Comment (iiLubos) [reply app/qml/inputs/MMTextBubblesEditor.qml:85]: unfortunately no, I spent a longer time solving this odd-length issue

File: app/qml/inputs/MMTextBubblesEditor.qml (hunk 4 above)
Comment (iiLubos) [reply app/qml/inputs/MMTextBubblesEditor.qml:109]: no, due to an odd behavior of the invisibleIds's length

File: app/qml/components/MMLinkedFeaturesDrawer.qml (hunk 8 above)
Comment (tomasMizera) [reply app/qml/components/MMLinkedFeaturesDrawer.qml:15]: Great! :) Thanks

**Comment by PeterPetrik** [comment]
please rebase

**Comment by iiLubos** [comment]
Ready to merge

**Review by tomasMizera — APPROVED** [review]

**Pull request merged**